### PR TITLE
add distributed RLHF based on Ray

### PR DESCRIPTION
**Pull request by wuxibin89**

### CODE REVIEW

**Review by catqaq — COMMENTED** [review]
#TODO(@wuxibin): save model
Is there anything special about saving models in the Ray environment?

**Comment by wuxibin89** [comment]
@catqaq I push this PR for starting reviewing earlier, I will implement save model soon.

File: examples/train_ppo_ray.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    actor_model = PPORayActorGroup(args.actor_num_nodes, args.actor_num_gpus_per_node, ActorModelRayActor)
+    reward_model = PPORayActorGroup(args.ref_num_nodes, args.ref_num_gpus_per_node, RewardModelRayActor)
+    critic_model = PPORayActorGroup(args.critic_num_nodes, args.critic_num_gpus_per_node, CriticModelRayActor)
+
+    # init reference/reward/actor mdoel
+    refs = []
+    refs.extend(ref_model.async_init_model_from_pretrained(strategy, args.pretrain, args.sft_model_path))
+    refs.extend(actor_model.async_init_model_from_pretrained(strategy, args.pretrain, args.sft_model_path))
+    refs.extend(reward_model.async_init_model_from_pretrained(strategy, args.critic_pretrain, args.reward_model_path))
+    ray.get(refs)
```
Comment (tsaoyu) [inline examples/train_ppo_ray.py:44]: Maybe it worthwhile to check how torchx drive the training steps. https://github.com/pytorch/torchx/blob/main/torchx/schedulers/ray/ray_driver.py#L232

A lot of `ray.get()` make it hard to understand where the actual Actor are runing.

File: openllama2/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -105,12 +115,25 @@ def __init__(
         else:
             self.kl_ctl = FixedKLController(init_kl_coef)
 
-        self.experience_maker = NaiveExperienceMaker(actor, critic, reward_model, initial_model, self.kl_ctl, strategy)
+        # if reference/reward/critic models are ray actor handle, we should use RemoteExperienceMaker.
+        if not isinstance(critic, ray.actor.ActorHandle):
```
Comment (tsaoyu) [inline openllama2/trainer/ppo_trainer.py:119]: Should we also check ref. and reward model here?

File: openllama2/trainer/ppo_utils/experience_maker.py [hunk 3]
```diff
@@ -186,3 +188,76 @@ def get_advantages_and_returns(
...
+            returns,
+            advantage,
+            attention_mask,
+            action_mask,
+            info,
+        )
+
+        # send experience to critic
+        experience_cpu = deepcopy(experience)
+        experience_cpu.to_device("cpu")
```
Comment (tsaoyu) [inline openllama2/trainer/ppo_utils/experience_maker.py:260]: It is kind pity to see due to limitation of Ray, we still have to move all tensors to CPU before calling `.remote()` funciton

File: openllama2/trainer/ray/launcher.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        self._master_addr = master_addr if master_addr else self._get_current_node_ip()
+        self._master_port = master_port if master_port else self._get_free_port()
+        os.environ["MASTER_ADDR"] = self._master_addr
+        os.environ["MASTER_PORT"] = str(self._master_port)
+        os.environ["WORLD_SIZE"] = str(self._world_size)
+        os.environ["RANK"] = str(self._rank)
+        # NOTE: Ray will automatically set the CUDA_VISIBLE_DEVICES
+        # environment variable for each actor, so always set device to 0
+        # os.environ["LOCAL_RANK"] = str(self._local_rank)
+        os.environ["LOCAL_RANK"] = "0"
```
Comment (tsaoyu) [inline openllama2/trainer/ray/launcher.py:34]: Will this conflict with deepspeed? DS also manage `CUDA_VISIBLE_DEVICES`

Comment (wuxibin89) [reply openllama2/trainer/ray/launcher.py:34]: DeepSpeed launcher is a thin wrapper of torchrun, we do not use it. Instead we manually spawn a group of processes as ray.Actor and assign rank for each of them (just like torchrun does), so there's no conflict here.

File: openllama2/trainer/ppo_trainer.py (hunk 2 above)
Comment (wuxibin89) [reply openllama2/trainer/ppo_trainer.py:119]: Well, it's more safe to check ref/rm, but check critic is enough here since ActorPPOTrainer should set ref/rm/critic all as ray.actor.ActorHandle.

File: examples/train_ppo_ray.py (hunk 1 above)
Comment (wuxibin89) [reply examples/train_ppo_ray.py:44]: `ray.get(refs)` here is a barrier to ensure all 4 models init correctly. I dig torchx for a while, actually it use ray.get as well :)
https://github.com/pytorch/torchx/blob/main/torchx/schedulers/ray/ray_driver.py#L241

**Comment by wuxibin89** [comment]
For anyone who is interested, additional training info
- wandb: [ppo_0917T11:07](https://wandb.ai/wuxibin/openllama2_train_ppo/runs/9dl5stpp?workspace=user-wuxibin)
- ppo checkpoint: [wuxibin/openllama2](https://huggingface.co/wuxibin/openllama2)

SFT answer
```
Please enter a prompt (or type 'exit' to quit): how do you think about Huawei?
1. Huawei is a Chinese multinational technology company that specializes in telecommunications, consumer electronics, and smart devices.

2. The company has made significant strides in the smartphone market, particularly in the mid-to-high-end segment, with its flagship lineup, the Huawei P series and Mate series.

3. Huawei has faced challenges due to the US trade ban, which has affected its ability to access crucial components and technologies from American suppliers. This has led to a decline in its smartphone sales and market share.

4. Despite these challenges, Huawei has continued to innovate and develop new technologies, such as its Kirin processors, which have been praised for their performance and efficiency.

5. The company has also made significant investments in research and development, focusing on areas like artificial intelligence, 5G, and smart city solutions.

6. Huawei has a strong presence in the global market, particularly in Asia, Europe, and Africa, and is considered one of the top smartphone manufacturers worldwide.

7. The company's commitment to innovation and customer satisfaction has led to a loyal fan base and a strong reputation in the tech industry.

8. However, due to the ongoing trade ban and the impact on its business, Huawei's future growth and market position remain uncertain.

In summary, Huawei is a major player in the smartphone market and has made significant strides in innovation and technology. However, the ongoing US trade ban has significantly impacted the company's growth and market share.
```
PPO answer
```
Please enter a prompt (or type 'exit' to quit): how do you think about Huawei?
1. Innovation: Huawei has made significant strides in the smartphone market, particularly in the areas of camera technology, battery life, and design. Their devices often offer innovative features and advanced hardware, such as triple-lens camera systems, large batteries, and sleek designs.

2. Competition: Huawei has been able to compete with Apple and other major smartphone manufacturers by offering high-quality, affordable devices. Their products have gained popularity in markets like China and Europe, where they have managed to capture a significant share of the smartphone market.

3. Security concerns: Huawei has faced some controversies and concerns regarding its ties to the Chinese government and potential security risks. This has led to restrictions on the use of Huawei devices in certain countries and has impacted its ability to access certain markets.

4. Research and Development: Despite these challenges, Huawei has continued to invest in research and development, focusing on areas such as artificial intelligence, 5G technology, and smart city solutions. This demonstrates their commitment to staying at the forefront of technological advancements.

5. Future prospects: As the smartphone market continues to evolve and new technologies emerge, Huawei has the potential to remain a major player in the industry. The company's focus on innovation, affordability, and global expansion could help them maintain their position in the market.

In conclusion, while Huawei has faced some challenges, it remains a significant player in the smartphone market, offering innovative products and competing effectively against established brands. The future of Huawei will depend on how they navigate the ongoing security concerns and continue to invest in research and development.
```

File: openllama2/trainer/ray/launcher.py (hunk 4 above)
Comment (tsaoyu) [reply openllama2/trainer/ray/launcher.py:34]: sure

File: openllama2/trainer/ppo_utils/experience_maker.py (hunk 3 above)
Comment (jovany-wang) [reply openllama2/trainer/ppo_utils/experience_maker.py:260]: Maybe it's worth to wrap a GPU buffer, but it's not in this PR's goal.

**Review by jovany-wang — APPROVED** [review]
As an initial POC of the implementation, it looks good.
